File: January2021/C++/Merge Sorted Array.cpp
/**
 * @author ssrbazpur
 * @Time Complexity: O(M+N) | Space:O(1)
 */

class Solution
{
public:
    void merge(vector<int> &nums1, int m, vector<int> &nums2, int n)
    {

        int p1 = m - 1;
        int p2 = n - 1;

        int track = nums1.size() - 1;

        while (track >= 0)
        {

            int firstVal = p1 >= 0 ? nums1[p1] : INT_MIN;
            int secondVal = p2 >= 0 ? nums2[p2] : INT_MIN;

            if (firstVal > secondVal)
            {
                nums1[track] = firstVal;
                p1--;
                track--;
            }
            else
            {
                nums1[track] = secondVal;
                p2--;
                track--;
            }
        }
    }
};

File: June2022/Java/MergeSortedArray.java

// @saorav21994
// TC : O(m+n)
// SC : O(m+n)

class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        int i = m-1;
        int j = n-1;
        int k = m+n-1;
        while (k >= 0) {
            if (j < 0 || (i >= 0 && nums1[i] > nums2[j]))
                nums1[k--] = nums1[i--];
            else 
                nums1[k--] = nums2[j--];
        }
        
    }
}
